### DISCUSSION TITLE
How to surround selected text with the typed character instead of replacing the text?

### DISCUSSION THREAD

**Discussion by rfgamaral** [ueberdosis/tiptap#2146] [General Questions] [answered]
I was trying to build an extension that would allow me to do exactly what GitHub does, something like this:

https://user-images.githubusercontent.com/96476/140745867-ec7e1893-8b6d-4bf3-84d2-0d2c2340bb6e.mp4

Funny enough, this doesn't seem to work on the new discussion input field, but works on issues 😅

Any pointers on how could I write an extension to do something like this? Not sure where to start for this one.

**Comment by philippkuehn** [comment]
I think you could implement a `wrapText()` command or something 🤔

**Reply by rfgamaral** [reply]
I guess what I'm trying to understand is where and how would I implement the trigger when the user types, for instance `*`, with text selected, so that I could replace the selected text with the typed character. I'm not sure where in the extension system would I go about writing all that.

**Comment by rfgamaral** [comment, marked as answer]
Okay, here's my attempt at implementing an extension that works very similar to GitHub ([CodeSandbox](https://codesandbox.io/s/extension-wrapselectionwith-klp3x)):

```js
const WRAPPING_SYMBOLS = {
    '*': null,
    _: null,
    '~': null,
    '"': null,
    "'": null,
    '`': null,
    '(': ')',
    '[': ']',
    '{': '}',
}

const WrapSelectionWith = Extension.create({
    name: 'wrapSelectionWith',
    addProseMirrorPlugins() {
        return [
            new Plugin({
                key: new PluginKey('wrapSelectionWith'),
                props: {
                    handleTextInput(view, from, to, text) {
                        const { state, dispatch } = view
                        const { selection, tr } = state

                        // Do not handle text input if selection is empty
                        if (selection.empty) {
                            return false
                        }

                        const wrappingSymbol = WRAPPING_SYMBOLS[text]

                        // Do not handle text input if no wrapping symbol was typed
                        if (wrappingSymbol === undefined) {
                            return false
                        }

                        // Insert wrapping symbols around the selected text while preserving hard
                        // line breaks and the selection itself
                        dispatch(
                            tr
                                .insertText(text, from, from)
                                .insertText(wrappingSymbol ? wrappingSymbol : text, to + 1, to + 1)
                                .setSelection(TextSelection.create(tr.doc, from + 1, to + 1)),
                        )

                        return true
                    },
                },
            }),
        ]
    },
})
```

Please note that this extension is meant to be used with some sort of "plain-text mode", and not fully WYSIWYG editor (it's not prepared to handle block elements, nor it converts them to the respective nodes). This "plain-text mode" is implemented in the CodeSandbox above, forcing a single top-level paragraph for the `Document`, and allowing only `Text` and `HardBreak` nodes.

---

@philippkuehn Would love your thoughts on this. If you find room for improvement, let me know.

**Reply by philippkuehn** [reply]
This looks absolutely fine 👍

**Reply by duhaime** [reply]
@rfgamaral did your sandbox move? The link you posted doesn't resolve anymore

**Reply by rfgamaral** [reply]
@duhaime I had deleted it, just recovered the link, try again please. But I don't recommend the "plain-text" approach you see there (i.e. single Paragraph, multiple hardBreaks). Instead, the document schema should allow for multiple paragraphs as the child nodes of the Document, and then Text, then tweak as necessary. This provides with a more consistent and expected editor behavior (we have too many workarounds for small issues the "plain-text" approach you see in the sandbox has caused, and I'm in the process of re-writing the whole thing).
